In [1]:
!pip install -q git+https://github.com/zhouy185/rosa-starter.git

In [ ]:
import numpy as np
from starter import ZONES, TIME_BLOCKS, COSTS, PROMISE, delivery_times

SEED = 1 # Fix a seed number so the simulation gives fixed numbers every time.

In [ ]:
# Print out the test so everything is good to go.
times = delivery_times("Far West", "Fri/Sat eve", 45, seed = SEED)
print(len(times), "orders; average delivery time", round(times.mean(), 1),
"minutes")

210 orders; average delivery time 42.9 minutes


# Part 1 Section A

In [ ]:
print(ZONES)
print(TIME_BLOCKS)
print(COSTS)
print(PROMISE)

['Central', 'North', 'Far West']
['Lunch', 'Weekday eve', 'Fri/Sat eve', 'Other']
{'refund': 10.0, 'churn_orders': 1.8, 'margin': 9.0}
45


In [ ]:
def get_times(promise, zone, time_block):

  # This function returns delivery times of all orders in the chosen zones and time block


  # If the user wants all zones or all time blocks, we need to handle that here.
  
  if zone == 'all':
    zones = ZONES
  else:
    zones = [zone]
  if time_block == 'all':
    blocks = TIME_BLOCKS
  else:
    blocks = [time_block]

  # delivery_times handles one zone and one block at a time, so loop over each pair
  arrays = []
  for z in zones:
    for b in blocks:
      arrays.append(delivery_times(z,b, promise, seed = SEED))

  # Put every order time into one array
  return np.concatenate(arrays)

In [ ]:
get_times(45, 'North', 'Lunch') # Check that the get_times function works properly

array([31.3, 36.1, 31.1, 19.1, 37. , 32.2, 24. , 33.6, 31.5, 30.8, 28.4,
       33.2, 22.6, 26.9, 24.4, 33.7, 28.5, 25.8, 22.3, 26.1, 28.3, 26. ,
       41.6, 38.1, 12.5, 16. , 26.8, 24.8, 30.1, 30.1, 53.2, 20.2, 25.2,
       52. , 34.2, 34.4, 24.2, 17.2, 29.7, 29.1, 19.5, 23. , 27.6, 21.2,
       27.4, 29. , 28.5, 24.2, 33.7, 36.8, 31. , 22.1, 35.1, 24.3, 36.7,
       20.4, 37.1, 28. , 19.4, 25.7, 28.7, 30.6, 21. , 20.2, 29.9, 24.5,
       30.3, 35.4, 17.2, 30.4, 40.7, 25.8, 22.1, 35.3, 30.4, 36.9, 25.4,
       18.1, 27.3, 24.7, 35.6, 29.9, 17.3, 19.7, 36.8, 34.6, 23.3, 28.2,
       32.2, 32.5, 36.7, 30.5, 27.4, 26.1, 38.7, 14.4, 27.1, 28.5, 18.4,
       31.2, 23.2, 36.5, 27.2, 34.5, 40.6, 31.6, 21.7, 17.9, 47.7, 27.3,
       22.9, 29.4, 26.6, 36.4, 28.5, 28.3, 22.8, 32.5, 20.7, 34.4, 44.5,
       17.8, 13.5, 33.9, 60.6, 20.9, 19.4, 33.6, 21.9, 24.2, 25.4, 33.1,
       25. , 30.7, 26.7, 21.9, 25.6, 21.2, 28.3, 20.1, 20.3, 43.7, 27.8,
       27.7, 32.9, 24.9, 26.3, 32. , 30.7, 19.9, 36

In [ ]:
len(get_times(45, 'North', 'Lunch')) # Check that the number of orders is correct

190

In [ ]:
def late_rate(promise, zone,time_block):
  # This function returns the percentage of orders delivered later than the promise.

  t = get_times(promise, zone, time_block) # Get all the delivery times for the chosen zone and time block
  late_count = np.sum(t > promise)   # how many are late
  total = len(t)                     # how many orders in total
  return round((100 * late_count / total), 2)

In [ ]:
p = 45  # promised delivery time in minutes

# Print out the late rates for different zones and time blocks to check that the late_rate function works properly.
print(f"Late rate, Far West / Fri/Sat eve (specific zone, specific block), {p}-minute promise: {round(late_rate(p, 'Far West', 'Fri/Sat eve'), 2)}% of orders were late\n")
print(f"Late rate, all zones / Lunch (all zones, specific block), {p}-minute promise: {round(late_rate(p, 'all', 'Lunch'), 2)}% of orders were late\n")
print(f"Late rate, North / all time blocks (specific zone, all blocks), {p}-minute promise: {round(late_rate(p, 'North', 'all'), 2)}% of orders were late\n")
print(f"Late rate, all zones / all time blocks (everything), {p}-minute promise: {round(late_rate(p, 'all', 'all'), 2)}% of orders were late\n")

Late rate, Far West / Fri/Sat eve (specific zone, specific block), 45-minute promise: 36.67% of orders were late

Late rate, all zones / Lunch (all zones, specific block), 45-minute promise: 3.33% of orders were late

Late rate, North / all time blocks (specific zone, all blocks), 45-minute promise: 6.8% of orders were late

Late rate, all zones / all time blocks (everything), 45-minute promise: 8.29% of orders were late



# Part 1 Section B

In [ ]:
results = []

# Loop over all zones and time blocks, and calculate the late rate for each combination. Store the results in a list of tuples.
for z in ZONES:
  for b in TIME_BLOCKS:
    results.append((z, b, late_rate(PROMISE, z, b)))


# Sort the results by late rate in descending order, so the highest late rates are at the top.
results.sort(key = lambda x: x[2], reverse= True)


# Print out the results in a nice format, with the rank, zone, time block, and late rate.
print(f'Ranking by late rate, {PROMISE}-minute promise\n')

rank = 1
for i in results:

  print(f"{rank}. {i[0]}, {i[1]}: {i[2]}%\n")
  rank += 1


Ranking by late rate, 45-minute promise

1. Far West, Fri/Sat eve: 36.67%

2. North, Fri/Sat eve: 16.67%

3. Far West, Weekday eve: 11.18%

4. Central, Fri/Sat eve: 6.84%

5. North, Weekday eve: 6.0%

6. Far West, Other: 4.69%

7. Central, Weekday eve: 4.24%

8. North, Lunch: 3.68%

9. Central, Lunch: 3.46%

10. Central, Other: 2.63%

11. North, Other: 2.5%

12. Far West, Lunch: 2.5%



# Part 1 Section C

In [ ]:
def avg_time(promise, zone, time_block):
  # This function returns the average delivery time (minutes) of orders in the chosen group.
  t = get_times(promise, zone, time_block)
  return round(np.mean(t), 2)

In [ ]:
p = 45  # promised delivery time in minutes

# Four cases: specific/specific, all/specific, specific/all, all/all
print(f"Average delivery time, Far West / Fri/Sat eve (specific zone, specific block), {p}-minute promise: {round(avg_time(p, 'Far West', 'Fri/Sat eve'), 2)} minutes\n")
print(f"Average delivery time, all zones / Lunch (all zones, specific block), {p}-minute promise: {round(avg_time(p, 'all', 'Lunch'), 2)} minutes\n")
print(f"Average delivery time, North / all time blocks (specific zone, all blocks), {p}-minute promise: {round(avg_time(p, 'North', 'all'), 2)} minutes\n")
print(f"Average delivery time, all zones / all time blocks (everything), {p}-minute promise: {round(avg_time(p, 'all', 'all'), 2)} minutes\n")

Average delivery time, Far West / Fri/Sat eve (specific zone, specific block), 45-minute promise: 42.92 minutes

Average delivery time, all zones / Lunch (all zones, specific block), 45-minute promise: 28.22 minutes

Average delivery time, North / all time blocks (specific zone, all blocks), 45-minute promise: 30.98 minutes

Average delivery time, all zones / all time blocks (everything), 45-minute promise: 31.19 minutes



# Part 1 Section D

In [ ]:
results_avg = []

# Loop over all zones and time blocks, and calculate the average delivery time for each combination. Store the results in a list of tuples.
for z in ZONES:
  for b in TIME_BLOCKS:
    results_avg.append((z, b, avg_time(PROMISE, z, b)))

# Sort by average time (3rd item of each tuple), slowest first
results_avg.sort(key = lambda x: x[2], reverse = True)

# Print out the results in a nice format, with the rank, zone, time block, and average delivery time.
print(f'Ranking by average delivery time, {PROMISE}-minute promise\n')

rank = 1
for i in results_avg:
  print(f"{rank}. {i[0]}, {i[1]}: {i[2]} minutes\n")
  rank += 1

Ranking by average delivery time, 45-minute promise

1. Far West, Fri/Sat eve: 42.92 minutes

2. North, Fri/Sat eve: 36.77 minutes

3. Far West, Weekday eve: 35.04 minutes

4. Central, Fri/Sat eve: 32.13 minutes

5. North, Weekday eve: 31.38 minutes

6. Far West, Lunch: 30.58 minutes

7. Far West, Other: 30.04 minutes

8. Central, Weekday eve: 29.03 minutes

9. North, Lunch: 28.67 minutes

10. North, Other: 27.66 minutes

11. Central, Lunch: 26.8 minutes

12. Central, Other: 26.04 minutes



# Part 1 Section E

The late rate is more relevant. Rosa's costs come from late orders, which triggers a refund and causes the customer to place fewer orders in the future. She does not pay anything for a slow average, so the late rate tracks her actual costs while the average delivery time does not.
<br><br>

The results from (b) and (d) show how the average can mislead. Far West on Fri/Sat evenings has an average delivery time of 42.92 minutes, which is under the 45-minute promise, so it looks acceptable. Yet 36.67% of its orders are late, the worst rate of any pair. The average hides this because a group can have many on-time orders pulling the average down while a large share still arrive late.
<br><br>

The rankings also disagree in places. Far West Lunch ranks 6th by average time (30.58 minutes) but is tied for last by late rate (2.5%), and it has a higher average than Central Weekday eve (29.03 minutes) despite a lower late rate (2.5% vs 4.24%). Ranking by average time would send Rosa's attention to the wrong places.
<br><br>

Average time still has some use as context, since the slowest pairs are generally the ones with the most late orders. But for deciding where to act and what a promise will cost, the late rate is the better measure.

# Part 2 Section A

In [ ]:
print(COSTS)

{'refund': 10.0, 'churn_orders': 1.8, 'margin': 9.0}


In [ ]:
def get_cost(costs):
  # This function returns the total cost of one late order
  cost = costs['refund'] + (costs['churn_orders'] * costs['margin']) # Calculate the total cost of one late order by looking up the refund, churn_orders, and margin values in the costs dictionary and applying the formula.
  return cost

In [ ]:
# Quick example to see of the function works properly.
print(f"Cost per late order: ${round(get_cost(COSTS), 2)}")

Cost per late order: $26.2


# Part 2 Section B

**Choosing the range of promises**

I will test promises from **20 to 80 minutes in steps of 5**.

- **Lower bound (20 minutes):** A promise below 20 minutes is unrealistic for food delivery. Almost every order would be late, so late costs would be at their highest and net profit would be strongly negative.
- **Upper bound (80 minutes):** I expect 80 minutes to be about the longest wait a customer would accept for delivery. Beyond that, demand would drop too far for the promise to be worth considering.
- **Step size (5 minutes):** A 5-minute step is fine enough to locate the promise with the highest net profit, without testing an excessive number of values.

After running the search, I will check that the best promise is not the first or last value in this range. If it is, the range would need to be extended.

In [ ]:
promises = list(range(20,81,5))

In [ ]:
# Now we need a function that will return the promise with the highest profit margin and the promise number

In [ ]:
def best_promise(zone, time_block, promises, costs):

  # This function returns the promise with the highest profit margin and the promise number
  late_cost = get_cost(costs)
  best_prom = None
  best_profit = None

  # Loop over all promises and calculate the profit for each one. Keep track of the best promise and best profit.
  for p in promises:
    t = get_times(p, zone, time_block)
    late_count = np.sum(t > p)
    total = len(t)
    profit = total * costs['margin'] - late_count * late_cost

    if best_profit is None or profit > best_profit:
      best_profit = profit
      best_prom = p

  return best_prom, best_profit
#

In [ ]:
# Quick example to see if the function works properly and returns the best promise and profit for a specific zone and time block.
best_prom, profit = best_promise("Far West", "Fri/Sat eve", promises, COSTS)
print(f"Far West / Fri/Sat eve: best promise = {best_prom} minutes, net profit = ${round(profit, 2)}")

Far West / Fri/Sat eve: best promise = 55 minutes, net profit = $1212.4


## AI Appendix

I used AI (Claude) in two ways: in a chat conversation while working on the notebook (Parts I and II), and in Claude Code while building the web app (Part III).

### 1. Claude chat: understanding the assignment and the notebook

- **Understanding the assignment.** I uploaded the assignment PDF and asked Claude to summarize it, explain what each part required, and point out key details in each section (for example, pooling orders for the `'all'` cases instead of averaging percentages, and checking that the best promise is not at the edge of the range).
- **Understanding the code.** I asked Claude to explain parts of the code I did not understand, such as the `get_times` helper, how `late_rate` works, and how the ranking sort with `lambda` works. I also asked for alternative ways to write the same logic.
- **Debugging and polish.** Claude helped me fix syntax errors, format the print statements with f-strings.


**Key prompts:**
1. "Observe the document carefully and explain what I need to do."
2. "Explain what I need to do in Part I before coding."
3. "Explain the code in cell 3." (the `get_times` helper)
4. "Can I write this in other ways?" (the `zones` / `blocks` if/else lines)
5. "Explain how the late rate code should works."
7. "What do the values in COSTS mean?"

### 2. Claude Code: building the Streamlit app (Part III)

I created a project-specific skill named `notebook-to-streamlit` (saved at `.github/skills/notebook-to-streamlit/SKILL.md`) and explicitly asked Claude Code to use that skill to port the logic from my notebook into the app. I asked Claude Code to create a Streamlit app to help Rosa decide the best promised delivery time, and specified that the app should have certain features.


**Key prompt:**
1. Read .github/skills/notebook-to-streamlit/SKILL.md and use that skill to build a Streamlit app from my notebook.
2. Select a zone and a time block from dropdown menus
3. Set the range of promised times to try
4. Adjust the profit margin per order, the estimated churn per late order, and the refund cost per late order
5. Display the recommended promised time for the selected zone and time block after clicking a button
6. Make the frontend look nice and professional, with a title, instructions, and a footer (dark theme would be preferable).
7. Make sure you include a table of the results for all promises in the selected range, sorted by net profit, with the best promise highlighted.

### 3. What I checked myself

I ran all notebook code and compared the outputs with my expectations. For the app, I tested it locally and confirmed it gives the same recommendation as my notebook for the same inputs before deploying.

## Github Link For My Repository off the Steamlit App

https://github.com/alperdalyazici/rosa-delivery-app.git

## The Link for My deployed app on Streamlit Community Cloud

https://rosa-delivery-app-yemwzrzy6plpfvenjuuh5k.streamlit.app/